File: 2015/s3_symmetric_cross.py


r, c = [int(i) for i in input().split()]
m = []
for _ in range(r):
    m.append([int(i) for i in input().split()])

res_size = -1
res_i = -1
res_j = -1

for i in range(1, r-1):
    for j in range(1, c-1):
        topHeight = i
        bottomHeight = r-i-1
        leftLength = j
        rightLength = c-j-1

        maxLength = min(min(topHeight, bottomHeight), min(leftLength, rightLength))

        size = 0

        for k in range(maxLength+1):
            if m[i][j - k] != m[i - k][j]:
                break

            if m[i - k][j] != m[i][j + k]:
                break

            if m[i][j + k] != m[i + k][j]:
                break

            if m[i + k][j] != m[i][j - k]:
                break
        
            size += 1

        if size > res_size:
            res_size = size
            res_i = i
            res_j = j

print(f"{res_size-1} {res_i+1} {res_j+1}")

File: 2015/s3_symmetric_cross.cpp


#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int main() {
    int r, c;
    cin >> r >> c;

    vector<vector<int>> m(r, vector<int>(c));

    for (int i = 0; i < r; i++) {
        for (int j = 0; j < c; j++) {
            cin >> m[i][j];
        }
    }

    int res_size = -1, res_i = -1, res_j = -1;

    for (int i = 1; i < r-1; i++) {
        for (int j = 1; j < c-1; j++) {
            int topHeight = i, bottomHeight = r-i-1, leftLength = j, rightLength = c-j-1;
            int maxLength = min(min(topHeight, bottomHeight), min(leftLength, rightLength));
            int size = 0;

            for (int k = 0; k <= maxLength; k++) {
                if (m[i][j-k] != m[i-k][j]) {
                    break;
                }

                if (m[i-k][j] != m[i][j+k]) {
                    break;
                }

                if (m[i][j+k] != m[i+k][j]) {
                    break;
                }

                if (m[i+k][j] != m[i][j-k]) {
                    break;
                }

                size += 1;
            }

            if (size > res_size) {
                res_size = size;
                res_i = i;
                res_j = j;
            }
        }
    }

    cout << res_size-1 << " " << res_i+1 << " " << res_j+1 << endl;
    return 0;
}